hal/timer/driver.go

```diff
@@ -4,61 +4,32 @@
 
 package timer
 
-import (
-	"github.com/embeddedgo/kendryte/p/bus"
-	"github.com/embeddedgo/kendryte/p/timer"
-)
-
-type Driver struct {
-	p *Periph
-
-	channel int
-}
-
-func NewDriver(p *Periph, ch int) *Driver {
-	d := Driver{
-		p:       p,
-		channel: ch,
-	}
-
-	return &d
+type DriverPWM struct {
+	*Channel
 }
 
-func (d *Driver) EnablePWM() {
-	d.p.EnableClock()
-
-	bits := timer.PWM_ENABLE | timer.ENABLE | timer.USER | timer.INTERRUPT
-	d.p.CH[d.channel].CONTROL.SetBits(bits)
+func PWM(ch *Channel) *DriverPWM {
+	return &DriverPWM{Channel: ch}
 }
 
-func (d *Driver) EnableISR() {
-	d.p.EnableClock()
-
-	// Avoid interrupt storm if frequency has not been set
-	if d.p.CH[d.channel].LOAD.Load() == 0 || d.p.LOAD_COUNT2[d.channel].Load() == 0 {
-		d.SetFrequency(1, .5)
-	}
-
-	// Clear any existing ISRs
-	d.p.ResetISR(d.channel)
+func (d *DriverPWM) Enable() {
+	d.Periph().EnableClock()
 
-	// Enable timer in user mode, unset interrupt mask if it was set
-	d.p.CH[d.channel].CONTROL.SetBits(timer.ENABLE | timer.USER)
-	d.p.CH[d.channel].CONTROL.ClearBits(timer.INTERRUPT)
+	d.control.SetBits(PWM_ENABLE | ENABLE | USER | INTERRUPT)
 }
 
-func (d *Driver) SetFrequency(frequency int, duty float64) {
-	clk := bus.APB0.Clock()
+func (d *DriverPWM) SetFrequency(frequency float64, duty float64) {
```

Comment:
I'm unsure about this function. The common PWM use case is setting frequency once and next set the duty cycle multiple times, sometimes very fast (or even very very fast using DMA). Does K210 timers allow us use them this way? That is, have separate `SetFrequency`/`SetInterval` and `SetDuty` methods?

You can avoid `float64` for duty if `SetFrequency` will return max duty value. But as I said before we can try with flotats, time will tell. We can also have two kind of set duty methods: `SetDuty` and `SetDutyFloat`.

See for example [this method](https://github.com/ziutek/emgo/blob/5f24fadd6bb33916f25e013398463f39a921f715/egpath/src/stm32/hal/tim/pwm.go#L77). As I sad  before, don't follow this interface because it contains STM32 specific things and I'm not completely happy wit it. The point was to allow very fast operation (by exporting the "duty register") and even use DMA (by allow obtaining the address of register).

Reply:
I see what you're saying, but I had a go at this, load_count is the interval in clock cycles to set the output high or trigger an interrupt, when load_count2 is non zero, it is the number of cycles to drive the output low, both counters must decrement before the next value of load_count 1 or 2 is read. You can test this by setting a 10 second interval and then changing it to a 1kHz immediately, the timer will wait a full 10 seconds before it starts running at 1kHz.